File: Solutions/src/CoinChangeDP.java
// Time Complexity :O(mn)
// Space Complexity :O(mn)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :no


// Your code here along with comments explaining your approach
class CoinChangeDP {
    public int coinChange(int[] coins, int amount) {
        int dp[][]= new int[coins.length+1][amount+1];
        int m = dp.length;
        int n = dp[0].length;
        for (int i=0;i<m;i++){
            dp[i][0]=0;
        }
        for (int j=0;j<n;j++){
            dp[0][j]=9999;
        }

        for(int i=1;i<m;i++){
            for(int j=1;j<n;j++){
                if(j<coins[i-1]){
                    dp[i][j]=dp[i-1][j];
                }
                else{
                    dp[i][j]=Math.min(dp[i-1][j],dp[i][j-coins[i-1]]+1);
                }
            }
        }
        int result = dp[m-1][n-1];
        if(result>=9999) return -1;
        else{
            return result;
        }
    }

}

File: coinChangeDp.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
    
        if coins == None or len(coins)==0:
            return None
        dp = [[12 for j in range(amount+1)] for i in range(len(coins)+1)]
#         col1    
        for i in range(0,len(dp)):
            dp[i][0] = 0
#         row1
        for j in range(1, len(dp[0])):
            dp[0][j] = amount+1
        
        for i in range(1, len(dp)):
            for j in range(1, len(dp[0])):
                
                if j<coins[i-1]:
                    dp[i][j] = dp[i-1][j]
                else:
                    dp[i][j] = min(dp[i-1][j], dp[i][j-coins[i-1]]+1)
        
        result = dp[len(dp)-1][len(dp[0])-1]
        
        if result>=amount+1:
            return -1
        
        return result
            
# time complexity would be O(amount*no. of coints) -> O(nk)
#  Approach is to make a dp matrix and not have to compute for repeated subproblems.
            
        
        
